### ISSUE TITLE
Bring back `dagger do`?

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#4189]
## Problem

The transition to a multi-language model has greatly improved the Dagger developer experience in many ways. But in its current form it has introduced a regression: *there no longer is a standard way to discover and run pipelines from the CLI*.

| | Dagger engine 0.2 (Europa) | Dagger engine 0.3 (Cloak) |
| -- | -- | -- |
| Discover pipelines | `dagger do --list` | ❌ |
| Run a pipeline | `dagger do <NAME>` | ❌ |

### Overall design

The current plan is to implement API extensions, more specifically:

1. Ability to add custom pipelines (written in any language) to the graphql schema
2. Ability to introspect available pipelines from the schema
3. Ability to invoke these pipelines from the CLI without crafting a full graphql schema

The design for this is still in progress, and timing is uncertain for delivering it.

### Short-term solution

This is issue is causing more pain and confusion than we originally expected. Yes developers love writing pipelines in their own language. But losing the convenience of a standard CLI entrypoint is a steep price to pay - especially in the context of CI integration.

Given the level of pain, and uncertainty on the timing of extensions, we should look for short-term solutions that don't compromise the long-term design. So far no design for such short-term solution has been proposed.

**Comment by sipsma** [comment]
I think any solution here is going to have to involve one way or another deriving an API from source code. That will allow the targets and the args they accept to be known and exposed through the CLI.

From previous work on extensions, we already have that in Go+Python. NodeJS could be backfilled too.

The shortest path I see to getting something like this that doesn't compromise our longer term plans would be:
1. Backfill support for nodejs code-first schemas
1. Add the ability to derive a CLI interface from a graphql schema
1. Use that to implement `dagger do`
   * We can either automatically derive the runtime (i.e. the load+execute logic specific to each language) based on file extensions, add a flag that allows manual specification of the runtime, or something else

This would be missing the ability to easily invoke other extensions from your extension (and associated version control) and the ability to run "remote code" (i.e. an source code that's in git rather than local), which is why it's a smaller effort than "full blown extensions". I think if we're careful though work we put into this could all be setup for full blown extensions rather than throwaway though.

**Comment by shykes** [comment]
> I think any solution here is going to have to involve one way or another deriving an API from source code. That will allow the targets and the args they accept to be known and exposed through the CLI.

One possible shortcut would be to ship a `dagger.json` file that can be produced and consumed in a useful way *now*, before extensions existed, and can be migrated in a reasonable way once extensions exist.

I think it could be as simple as a:

```
// Dagger project file schema
name?: string
description?: string
// All types to be deprecated post-extensions, except "graphql"
type: "go"|"python"|"javascript"|"typescript"|"exec"|"graphql"
// Path of executable entrypoint, relative to project directory
// Only required for project type "exec"
entrypoint?: string
```

Then `dagger do` would use this file to list & invoke pipelines.

**Comment by sipsma** [comment]
> One possible shortcut would be to ship a dagger.json file that can be produced and consumed in a useful way now, before extensions existed, and can be migrated in a reasonable way once extensions exist.

I'm missing some pieces here. For a concrete example, say I wrote some go code like this:
```go
func Echo(input string) string {
  return input
}

func Yell(input string) string {
  return input + "!!!!!"
}
```

How would I invoke the targets using `dagger.json` and `dagger do` without `dagger` doing some magic to derive an API from that source code?

**Comment by shykes** [comment]
> > One possible shortcut would be to ship a dagger.json file that can be produced and consumed in a useful way now, before extensions existed, and can be migrated in a reasonable way once extensions exist.
> 
> I'm missing some pieces here. For a concrete example, say I wrote some go code like this:
> 
> ```go
> func Echo(input string) string {
>   return input
> }
> 
> func Yell(input string) string {
>   return input + "!!!!!"
> }
> ```
> 
> How would I invoke the targets using `dagger.json` and `dagger do` without `dagger` doing some magic to derive an API from that source code?

I'm picturing a very simple convention with less magic than you imagine:

* Developer must write a standalone program (the entrypoint), and write it to...
  * `main.go` for `type: "go"`
  * `index.js` for `type: "js"`
  * `index.ts` for `type: "ts"`
  * `main.py` for `type: "python"`
  * `<value of entrypoint field>` for `type: "exec"`
* `dagger do` invokes the pipeline by executing the entrypoint, and passing the pipeline name as argument

**Comment by sipsma** [comment]
> Developer must write a standalone program (the entrypoint), and write it to... 

I see, so for go, users would be responsible for writing a main func that actually invokes targets based on a command line arg, right? If we implemented this, documented it for users and then they asked what the benefits of using `dagger do` over just writing the same main func and invoking it directly with `go run` are, what would our answer be? It seems like unless we add some kind of sugar it just becomes mostly boilerplate around the same interface you get from `go run`.

**Comment by shykes** [comment]
> > Developer must write a standalone program (the entrypoint), and write it to...
> 
> I see, so for go, users would be responsible for writing a main func that actually invokes targets based on a command line arg, right? If we implemented this, documented it for users and then they asked what the benefits of using `dagger do` over just writing the same main func and invoking it directly with `go run` are, what would our answer be? It seems like unless we add some kind of sugar it just becomes mostly boilerplate around the same interface you get from `go run`.

*Disclaimer: I'm just exploring this idea, so it may not actually solve anything. With that said...*

The benefit would be a standardized entrypoint, which in turn would mean a (more) standardized way of invoking Dagger from CI.

The motivation for this issue is users reporting that they are generally confused as to how they're supposed to use Dagger, and how it maps in their familiar CI experience.

[Example on Discord today](https://discord.com/channels/707636530424053791/708371226174685314/1052301428397457459) from @sw360cab: 

> Hi guys. Long time I am not joining here... I am stunned by all the updates/changes that are going on.
I started playing with Go SDK and I was wondering if there is a final goal to ever reach Dagger Cue SDK (formerly Dagger CLI), if it makes sense.
> I mean Dagger Cue SDK is really able to create portable Plans and to make them available throughout the various/major CI/CD providers. Will be something like that possible through Go SDK in the future or they are intended to remain separate entities with a different purpose?

> So how can I make a comparison between what I currently can achieve with Dagger CUE plan + Dagger Cue CLI installed on CI/CD provider (e.g. Github Actions) and what I can write down using an SDK (e.g. GO)?
> Because right now I was able to replicate my Plan in the SDK but I am struggling understanding what I can do with that, compared to how powerful is porting my plan to any CI/CD provider by just installing the Dagger CUE client into.
I hope it make sense what I am writing, Maybe I am mixing up concepts and just doing confusion and noise 😵‍💫

>  I felt a little bit lost, trying to connect the dots.

**Comment by marcosnils** [comment]
Reading this thread something popped in my head that I think shouldn't go unnoticed. If we bring back `dagger do` to improve the DX for discovering and running pipelines, we should be mindful about what other aspects of the DX we're hurting because of this. 

For example: 

- Writing boilerplate code so the CLI can discover and run pipelines. 
- Confusion between different pipeline entry-points. Some users might use `dagger do`, others `go run` for convenience because they don't want to install the CLI for example
- Adding an extra step to the CI setup pipeline to download the CLI

At this stage I'm not sure if there's something we could do that provides "the best" experience until we have the API extensions. Do you think this could be temporarily mitigated by improving our docs?

**Comment by shykes** [comment]
> * Writing boilerplate code so the CLI can discover and run pipelines.

This will be true with extensions.

> * Confusion between different pipeline entry-points. Some users might use `dagger do`, others `go run` for convenience because they don't want to install the CLI for example

This will also be true with extensions.

> * Adding an extra step to the CI setup pipeline to download the CLI

We should already recommend this, so you can configure your session with `dagger run`.

> At this stage I'm not sure if there's something we could do that provides "the best" experience until we have the API extensions. Do you think this could be temporarily mitigated by improving our docs?

I don't want to dismiss your concerns - I'm not even sure this particular design is a good idea. BUT I am beginning to worry that "the extensions will do this" is becoming an excuse to ignore the problem completely. "The extensions" are not going to design themselves, we need to keep the design process alive, and attempts at incremental improvements like this one are part of that.

**Comment by marcosnils** [comment]
> BUT I am beginning to worry that "the extensions will do this" is becoming an excuse to ignore the problem completely.

I totally understand your feelings and I'm not against of shipping something intermediate that can alleviate users pain while bringing us a bit closer to the final solution. Having said that, the main benefit of API extensions is that it absorbs the complexity of having to configure the SDK runtime, since that'll be provided by the engine. In my head, with extensions there's a clear API separation between the CLI (loader) and the pipeline code itself. Users will only need to install the `dagger` CLI and everything else will be automatically configured for them. 

It'd be nice if we can find an intermediate solution somehow that could bring us closer to that final experience. For example, is there something we can do in the engine so all SDK runtimes are provided to the user as containers and then `dagger do` somehow uses that without needing the runtime to be installed locally?  I know I'm not proposing anyting about how users will define and call their pipelines, but I mostly wanted to highlight the opportunity to start simplifying the runtime setup to the user as it'll happen with Extensions. 


Maybe the previous paragraph could be an improvement over your initial proposal Solomon. Keep a dagger.json (or similar) descriptor, but also enable the engine to be able to bootstrap the pipeline language runtime and start it. I'm not sure if this is what you had in mind in your initial post, my interpretation was that the runtime would still have to be present in the local machine where `dagger do` is triggered.

**Comment by sipsma** [comment]
> This would help make Dagger more understandable to users that are familiar with the "normal" way of doing CI, and are put off by the current situation. The motivation for this issue is users reporting that they are generally confused as to how they're supposed to use Dagger, and how it maps in their familiar CI experience.

Yeah that makes sense. I think if we pursue something like this it'd be best to really optimize for low boilerplate. The benefits of this first step are real but marginal, so even a small bit of extra overhead can make it overall not worth it IMO.

That makes me feel like this would be worth it if we could avoid dagger.json. Along those lines, thinking out loud:
1. What if instead of adding `dagger do` we add an extra feature to `dagger run` that enables you to do e.g. `dagger run ./main.go`, or, if you are in a nodejs project, `dagger run test` where `test` is a script in package.json.
2. This would require a lot of autodetection magic; probably too much honestly since it would be hard to distinguish between running a command called `test` and a package.json script called `test`.
3. I suppose we could add a flag like `dagger run --sdk nodejs test` but at that point why not just have users call `dagger run yarn test`

So yeah, I'm not sure if I see a path that makes this worth it yet. It's a choice between excessive boilerplate or excessive magic (the bad+confusing kind, not the delightful kind). But maybe there's a middle ground I'm not seeing yet.

---

I agree that docs might be the best short term option here; can just show commands like `dagger run go run ...` and `dagger run yarn ...`.

There is a grain of optimism in me that re-using the extension code-first schema support to expose a pipeline API via command line args might be easier than it sounds, mostly just because the hardest part of that work is already out of the way for go+python at least. If there's time in the near future to experiment a bit I may give it a shot (or anyone else should feel free to also)

**Comment by marcosnils** [comment]
> There is a grain of optimism in me that re-using the extension code-first schema support to expose a pipeline API via command line args might be easier than it sounds,

This would be a possible alternative to the "bring us closer to the final implementation" idea I was trying to transmit in my previous post.

**Comment by shykes** [comment]
This makes me want to hold off on un-hiding `dagger run`, to keep the door open to a better abstraction that combines `dagger do` and `dagger run` in a way that improves the experience today, and is forward-compatible with extensions.

**Comment by shykes** [comment]
> Users will only need to install the `dagger` CLI and everything else will be automatically configured for them.
> 
> It'd be nice if we can find an intermediate solution somehow that could bring us closer to that final experience.

Isn't that what we're discussing now? What does this "final experience" look like exactly from the command-line? We haven't actually defined that. Our extensions design so far defines 1) how to package pipeline logic into graphql resolvers that the engine can load and run; 2) how to invoke it from a graphql query. It doesn't define how end users will experience this from the CLI. So why not start defining it now?

Options I see so far:

1) Send GraphQL queries directly: `dagger query`
2) Execute a client tool directly: CLI experience undefined (depends on the tool)
3) Execute a client tool via `dagger run`: CLI experience partially undefined (arguments to `dagger run` depend on the tool)
4) Execute `dagger do STEP [...STEP]` where each STEP maps to a graphql query

My point I suppose, is that the problem of standardizing the CLI experience is related to extensions, but orthogonal. We have this problem now (sooner and stronger than I expected), and we will continue to have it with extensions. When we say "extensions will solve this" that's not technically true, what we're really saying is "why worry about this until we have extensions". The answer as it turns out is: "because users are confused about a lack of standardized CLI experience *today*".

**Comment by shykes** [comment]
> is there something we can do in the engine so all SDK runtimes are provided to the user as containers and then `dagger do` somehow uses that without needing the runtime to be installed locally? I know I'm not proposing anyting about how users will define and call their pipelines, but I mostly wanted to highlight the opportunity to start simplifying the runtime setup to the user as it'll happen with Extensions.

That would be nice, and definitely an additional OX improvement compared to what I proposed (no runtime dependencies on Go, NodeJS etc!), but it's a breaking change for the DX: right now your client tool has full access to the host, but that's no longer the case if we run it in a container. This breaking change is inevitable with extensions, but since it's something new and exciting, it's a good tradeoff. Different story if we break the behavior of your client tools without giving you any DX improvement in return..

**Comment by marcosnils** [comment]
> This breaking change is inevitable with extensions, but since it's something new and exciting, it's a good tradeoff. Different story if we break the behavior of your client tools without giving you any DX improvement in return..

Agree with your points about being a breaking change for DX and the Extensions particularity.

 
> My point I suppose, is that the problem of standardizing the CLI experience is related to extensions, but orthogonal. We have this problem now (sooner and stronger than I expected), and we will continue to have it with extensions. When we say "extensions will solve this" that's not technically true, what we're really saying is "why worry about this until we have extensions". The answer as it turns out is: "because users are confused about a lack of standardized CLI experience _today_".

I also agree with this but, is it really something we can do something better? The fact that dagger is designed to be used primarily as an SDK is a core and intended decision, and trying to come up with a `CLI` as some sort of "universal loader/runner" across all languages seems like something that doesn't fit in the original design which we're trying to force. This reminds me to the same thing that happened with containers when people tried to put fully bloated init systems and tried to use them as "lightweight VM's" (besides the fact that some people currently do it).

Would it hurt to start with better docs explaining the reason for this and then referencing how to setup CI pipeline using standard loaders for each language? (I think we have this examples today). Given the ideas that we have discussed so far, it doesn't seem to me that they are improving the overall DX and I'm running out of ideas. I'll keep a thread in my head for this.

**Comment by helderco** [comment]
Yes, I'm with @marcosnils generally I think. Speaking for Python here, there's many ways to run a python script. If there would be a unified entrypoint, to me it would be running in a container, but you lose direct access to host for example.

I also have some use cases in mind for Dagger that are not running from the terminal. So the simplicity of running a python script is good enough for me. Python CI pipelines already setup a Python environment usually. We just make that simpler. Still nice to "document" what we think is a good solution for declaring pipeline steps (like _lint_ and _test_). Either with conveniences from the SDK for less boilerplate, or just a guide.

You have a nice chainable API you can use, but it's not required. You could ditch the whole SDK, run `dagger listen` manually and use an entirely different GraphQL library, or raw query strings. Same thing with running from the terminal experience vs some other environment.

But there's different package managers, different python interpreters or even their name. So it's hard to see how this can be implemented in a simple way. Seems like added complexity that's not worth it.

Some of the confusion users are having may come from the previous version of Dagger where you ran `.cue` pipelines with the CLI. We just need to have clear documentation on this.

**Comment by shykes** [comment]
I'm going to take a step back here. Forget about `dagger do` or any hypothetical future design: I think there's confusion about the *current design* and the fact that it already defines a standardized CLI entrypoint: `dagger run`.

If you're asking "what would be the point of a standardized CLI entrypoint" I'm sorry to say that that ship has sailed: we already have one and will be relying on it more and more in the context of CI integration, because that is how we will offer the best OX. The only question is: should we do this with `dagger run` as designed currently, or can we further improve the experience with a higher-level `dagger do`? Not having a standardized CLI entrypoint at all, or keeping it as an exception instead of the recommended way to run things, is not an option.

**Comment by helderco** [comment]
Let me rephrase.

You can run today `dagger run python script.py` while not even using/importing the [Python SDK](https://pypi.org/project/dagger-io/).

The same way it's opt-in and gives you some conveniences for a nicer `script.py` (as opposed do doing raw queries), the SDK could also support an opinionated way to declare "steps" that can be detected by `dagger do`.

It's just another convenience layer, I'm ok with that.

**Comment by sagikazarmark** [comment]
If I may add a few comments from a user's perspective:

Most of my projects already have a Makefile or a Taskfile.yaml or something that acts a task runner for the project. (Despite how nice and easy `go build` and `go test` are they are just not good enough for all around everyday use)

I most certainly will continue having those tools around for development anyway. I can easily integrate my Dagger execution into them.

I'd much rather do that compared to using a temporary, short-term solution that may change significantly. I'm (and my peers are) already familiar with our current toolings, so integrating Dagger into those is potentially an even better way to introduce a new tool to the chain.

I'll also say that not having to download a CLI felt good. Even though I use Nix a lot (I had a bunch of practices to download CLI tools (eg. from Makefiles) even before that), not having to download a new CLI every other week to get the new features made upgrading to the Go SDK easier a lot.

It seems to me that this is mostly a problem for users who want to _use_ Dagger. I don't want to use Dagger, I want to run CI. That fact that it's Dagger under the hood is great, but I actually don't mind that people don't know/care what it is.

Obviously, someone has to write those pipelines, but once I'm working on that, I already know enough about Dagger to either use `go run` (and I don't really need discovery) or just run Dagger through my usual task runners (eg. `make ci`).

I don't want to diminish anyone's pain who feels the need for a CLI, I just wanted to say there are alternatives and I'll probably prefer sticking to those over a temporary solution.

---

To offer some ideas as well: I wonder if the SDKs could offer a way to provide insights into pipelines to the Dagger CLI. Thinking about the d2 graph we saw in the community call yesterday. I know those are generated from logs, but the information is still there, I'm just not sure how to extract it (eg. passing a dummy client to pipeline code that doesn't actually run anything just traces the different calls and builds the graph).

**Comment by shykes** [comment]
This has been solved with the [introduction of Dagger Functions](https://dagger.io/blog/introducing-dagger-functions), and `dagger call` :)
